x/lightclient/keeper/canonical_client.go

```diff
@@ -0,0 +1,122 @@
+	"github.com/dymensionxyz/dymension/v3/x/lightclient/types"
+	rollapptypes "github.com/dymensionxyz/dymension/v3/x/rollapp/types"
+)
+
+const (
+	ibcRevisionNumber = 1
+)
+
+// GetProspectiveCanonicalClient returns the client id of the first IBC client which can be set as the canonical client for the given rollapp.
+// The canonical client criteria are:
+// 1. The client must be a tendermint client.
+// 2. The client state must match the expected client params as configured by the module
+// 3. All the existing consensus states much match the corresponding height rollapp block descriptors
+func (k Keeper) GetProspectiveCanonicalClient(ctx sdk.Context, rollappId string, stateInfo *rollapptypes.StateInfo) (clientID string, stateCompatible bool) {
+	clients := k.ibcClientKeeper.GetAllGenesisClients(ctx)
+	for _, client := range clients {
+		clientState, err := ibcclienttypes.UnpackClientState(client.ClientState)
+		if err != nil {
+			continue
+		}
+		// Cast client state to tendermint client state - we need this to get the chain id and state height
+		tmClientState, ok := clientState.(*ibctm.ClientState)
+		if !ok {
+			continue
+		}
+		if tmClientState.ChainId != rollappId {
+			continue
+		}
+		if !types.IsCanonicalClientParamsValid(tmClientState) {
+			continue
+		}
+		sequencerPk, err := k.GetSequencerPubKey(ctx, stateInfo.Sequencer)
+		if err != nil {
+			continue
+		}
+		res, err := k.ibcClientKeeper.ConsensusStateHeights(ctx, &ibcclienttypes.QueryConsensusStateHeightsRequest{
+			ClientId:   client.GetClientId(),
+			Pagination: &query.PageRequest{Limit: stateInfo.GetLatestHeight()},
+		})
+		if err != nil {
+			continue
+		}
+		stateCompatible = true
+		for _, consensusHeight := range res.ConsensusStateHeights {
+			consensusState, _ := k.ibcClientKeeper.GetClientConsensusState(ctx, client.ClientId, consensusHeight)
+			tmConsensusState, _ := consensusState.(*ibctm.ConsensusState)
+			ibcState := types.IBCState{
+				Root:               tmConsensusState.GetRoot().GetHash(),
+				NextValidatorsHash: tmConsensusState.NextValidatorsHash,
+				Timestamp:          time.Unix(0, int64(tmConsensusState.GetTimestamp())),
+			}
+
+			var rollappState types.RollappState
+			bd, found := stateInfo.GetBlockDescriptor(consensusHeight.GetRevisionHeight())
+			if found {
+				rollappState.BlockDescriptor = bd
+				rollappState.NextBlockSequencer = sequencerPk
+			} else {
+				// Look up the state info for the block descriptor
+				oldStateInfo, err := k.rollappKeeper.FindStateInfoByHeight(ctx, rollappId, consensusHeight.GetRevisionHeight())
+				if err != nil {
+					stateCompatible = false
+					break
+				}
+				bd, _ = oldStateInfo.GetBlockDescriptor(consensusHeight.GetRevisionHeight())
+				oldSequencerPk, err := k.GetSequencerPubKey(ctx, oldStateInfo.Sequencer)
+				if err != nil {
+					stateCompatible = false
+					break
+				}
+				rollappState.BlockDescriptor = bd
+				rollappState.NextBlockSequencer = oldSequencerPk
```

Comment:
I'm finding it hard to reason about this. I can't understand where the +1 for the next validators hash comes into it